packages/perseus/src/widgets/number-line/number-line.tsx

```diff
@@ -144,36 +143,39 @@ const TickMarks: any = Graphie.createSimpleClass((graphie, props) => {
         base = undefined;
     }
 
-    // Draw and save the tick marks and tick labels
-    for (let i = 0; i <= props.numDivisions; i++) {
-        const x = range[0] + i * props.tickStep;
-        results.push(graphie.line([x, -0.2], [x, 0.2]));
-
-        const labelTicks = props.labelTicks;
-        if (labelTicks || props.labelStyle === "decimal ticks") {
-            if (x === leftLabel || x === rightLabel) {
-                results.push(
-                    graphie.style({color: KhanColors.BLUE}, () =>
-                        _label(graphie, props.labelStyle, x, x, base),
-                    ),
-                );
-            } else {
-                results.push(_label(graphie, props.labelStyle, x, x, base));
-            }
-        }
-    }
-
-    // Render the labels' lines
-    graphie.style(
-        {
-            stroke: KhanColors.BLUE,
-            strokeWidth: 3.5,
-        },
-        () => {
-            results.push(graphie.line([leftLabel, -0.2], [leftLabel, 0.2]));
-            results.push(graphie.line([rightLabel, -0.2], [rightLabel, 0.2]));
-        },
+    const endpointLineStyle = {
+        stroke: KhanColors.BLUE,
+        strokeWidth: 3.5,
+    };
+    const endpointTextStyle = {color: KhanColors.BLUE};
+
+    // Generate an array of tick numbers:
+    //    `Array(props.numDivisions)` makes an array of null values - one for every division marker
+    //    `.keys()` gets the index values for each marker placeholder
+    //    `.map()` converts the index values into actual tick numbers
+    const initialTicks: number[] = [...Array(props.numDivisions).keys()].map(
+        (index) => range[0] + index * tickStep,
     );
+    // Ensure that any label markers and range endpoints are included in the array
+    // Using `Set()` prevents duplication of tick numbers (and is quite performant)
+    const allTicks: number[] = [
+        ...new Set([...initialTicks, leftLabel, rightLabel, ...range]),
+    ].sort((current, next) => current - next);
+
+    // Cycle through each tick number and add a tick line, and a label (if needed)
+    allTicks.forEach((tick) => {
+        const tickHasLabel = tick === leftLabel || tick === rightLabel;
```

Comment:
Based on how we use this variable, I think `tickIsEndpoint` might be a clearer name.

- A tick might have a label for reasons other than `tickHasLabel` being true.
- When this variable is true, we use `endpointLineStyle` and `endpointTextStyle`.

```suggestion
        const tickIsEndpoint = tick === leftLabel || tick === rightLabel;
```

Reply:
Sounds good to me!